src/librustc_traits/lowering.rs

```diff
@@ -363,10 +365,60 @@ pub fn program_clauses_for_type_def<'a, 'tcx>(
 }
 
 pub fn program_clauses_for_associated_type_def<'a, 'tcx>(
-    _tcx: TyCtxt<'a, 'tcx, 'tcx>,
-    _item_id: DefId,
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    item_id: DefId,
 ) -> Clauses<'tcx> {
-    unimplemented!()
+    // Rule ProjectionEq-Normalize
+    //
+    // ProjectionEq can succeed by normalizing:
+    // ```
+    // forall<Self, P1..Pn, Pn+1..Pm, U> {
+    //   ProjectionEq(<Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> = U) :-
+    //       Normalize(<Self as Trait<P1..Pn>>::AssocType<Pn+1..Pm> -> U)
+    // }
+    // ```
+    let item = tcx.associated_item(item_id);
+    debug_assert_eq!(item.kind, ty::AssociatedKind::Type);
+    let trait_id = match item.container {
+        ty::AssociatedItemContainer::TraitContainer(trait_ref) => trait_ref,
+        _ => bug!("not an trait container"),
+    };
+
+    // `Self as Trait<P1..Pn>`
+    let trait_ref = tcx.impl_trait_ref(trait_id).unwrap();
+
+    // `U`
+    let ty = tcx.type_of(item_id);
```

Comment:
I don’t think this will work, what we want to do is to introduce a new type parameter `U`. Not sure how to do that (probably through `tcx.mk_ty_param`), I’ll try and get back to you for more guidance.

Reply:
Ok so as you can see in the description of the rule, we want the output to be something like:
```
ProjectionEq(<Self as Trait>::Item = U) :- Normalize(<Self as Trait>::Item = U)
```
where `U` is a new type variable. This could be done naively by having:
```rust
// add a type param at depth `0`
let ty = tcx.mk_ty_param(0, Symbol::intern(“U”).as_interned_str());
```
however as you can see, rustc refers to a type param with a depth index *and* an explicit name (here “U”), instead of referring to all parameters by using indices only. Hence if the name `U` is already taken, as in:
```
trait Foo<U> {
    #[rustc_dump_program_clauses]
    type Item;
}
```
then the output clause will be wrong as it will use the same parameter `U` for two unrelated things.

A possible fix would be to choose `U` whenever this name is not already in use, and in case it is used just choose a random name which is not already in use. Another possible fix would be two `Binder` layers instead of only one, as I think type parameters can have the same name if they are not at the same depth. I’ll confer with niko.

Reply:
Hmm so the problem here is that rustc doesn't really support this yet

Reply:
I wonder if we can leave this as a kind of FIXME for now